README.md
[![Build Status](https://travis-ci.org/s-aska/dropbox-api-command.svg?branch=master)](https://travis-ci.org/s-aska/dropbox-api-command)
# NAME

App::dropboxapi - command line interface to access Dropbox API

# SYNOPSIS

```
dropbox-api put /tmp/foo.txt dropbox:/Public/
```

Run `dropbox-api help` for more options.

# DESCRIPTION

dropbox-api is a command line interface to access Dropbox API.

- ls
- find
- du
- sync
- cp
- mv
- rm
- mkdir
- get
- put

# Install and Setup

## 1. Install

### 1-a) FreeBSD

```
pkg_add -r dropbox-api-command
```

### 1-b) Ubuntu

```
sudo apt-get install make gcc libssl-dev wget
wget https://raw.github.com/miyagawa/cpanminus/master/cpanm
sudo perl cpanm App::dropboxapi
```

### 1-c) CentOS

```
# CentOS 5
sudo yum install gcc gcc-c++ openssl-devel wget
# CentOS 6
sudo yum install gcc gcc-c++ openssl-devel wget perl-devel
wget https://raw.github.com/miyagawa/cpanminus/master/cpanm
sudo perl cpanm App::dropboxapi
```

### 1-d) OS X

```
# Install Command Line Tools for Xcode
open https://www.google.com/search?q=Command+Line+Tools+for+Xcode

curl -O https://raw.github.com/miyagawa/cpanminus/master/cpanm
sudo perl cpanm App::dropboxapi
```

## 2. Get API Key and API Secret

```perl
https://www.dropbox.com/developers
My Apps => Create an App
```

## 3. Get Access Token and Access Secret

```
> dropbox-api setup
Please Input API Key: ***************
Please Input API Secret: ***************
1. Open the Login URL: https://www.dropbox.com/oauth2/authorize?client_id=*****&response_type=code
2. Input code and press Enter: ***************
success! try
> dropbox-api ls
> dropbox-api find /
```

## 4. How to use Proxy

Please use -e option.

```
> HTTP_PROXY="http://127.0.0.1:8888" dropbox-api setup -e
```

# Sub Commands

## help

disp help.

- syntax

    dropbox-api help \[&lt;command>\]

### Example

```perl
> dropbox-api help
Usage: dropbox-api <command> [args] [options]

Available commands:
    setup get access_key and access_secret
    ls    list directory contents
    find  walk a file hierarchy
    du    disk usage statistics
    cp    copy file or directory
    mv    move file or directory
    mkdir make directory (Create intermediate directories as required)
    rm    remove file or directory (Attempt to remove the file hierarchy rooted in each file argument)
    put   upload file
    get   download file
    sync  sync directory (local => dropbox or dropbox => local)

Common Options
    -e enable env_proxy ( HTTP_PROXY, NO_PROXY )
    -D enable debug
    -v verbose
    -s sandbox mode, but this option has been removed.

See 'dropbox-api help <command>' for more information on a specific command.
```

### Example ( command help )

```
> dropbox-api help ls
Name
    dropbox-api-ls - list directory contents

SYNOPSIS
    dropbox-api ls <dropbox_path> [options]

Example
    dropbox-api ls Public
    dropbox-api ls Public -h
    dropbox-api ls Public -p "%d\t%s\t%TY/%Tm/%Td %TH:%TM:%TS\t%p\n"

Options
    -h print sizes in human readable format (e.g., 1K 234M 2G)
    -p print format.
        %d ... is_dir ( d: dir, -: file )
        %i ... id
        %n ... name
        %p ... path_display
        %P ... path_lower
        %b ... bytes
        %s ... size (e.g., 1K 234M 2G)
        %t ... server_modified
        %c ... client_modified
        %r ... rev
        %Tk ... DateTime 'strftime' function (server_modified)
        %Ck ... DateTime 'strftime' function (client_modified)
```

[http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime\_Patterns](http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime_Patterns)

## ls

file list view.

- alias

    list

- syntax

    dropbox-api ls &lt;dropbox\_path>

### Example

```
> dropbox-api list /product
d        - Thu, 24 Feb 2011 06:58:00 +0000 /product/chrome-extentions
-   294557 Sun, 26 Dec 2010 21:55:59 +0000 /product/ex.zip
```

### human readable option ( -h )

print sizes in human readable format (e.g., 1K 234M 2G)

```
> dropbox-api ls /product -h
d        - Thu, 24 Feb 2011 06:58:00 +0000 /product/chrome-extentions
-  287.7KB Sun, 26 Dec 2010 21:55:59 +0000 /product/ex.zip
```

### printf option ( -p )

print format.

```
> dropbox-api ls /product -p "%d\t%s\t%TY/%Tm/%Td %TH:%TM:%TS\t%p\n"
d       -       2011/02/24 06:58:00     /product/chrome-extentions
-       287.7KB 2010/12/26 21:55:59     /product/ex.zip

    %d ... is_dir ( d: dir, -: file )
    %i ... id
    %n ... name
    %p ... path_display
    %P ... path_lower
    %b ... bytes
    %s ... size (e.g., 1K 234M 2G)
    %t ... server_modified
    %c ... client_modified
    %r ... rev
    %Tk ... DateTime 'strftime' function (server_modified)
    %Ck ... DateTime 'strftime' function (client_modified)
```

[http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime\_Patterns](http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime_Patterns)

## find

recursive file list view.

- syntax

    dropbox-api find &lt;dropbox\_path> \[options\]

### Example

```
> dropbox-api find /product/google-tasks-checker-plus
/product/chrome-extentions/google-tasks-checker-plus/README.md
/product/chrome-extentions/google-tasks-checker-plus/src
/product/chrome-extentions/google-tasks-checker-plus/src/background.html
/product/chrome-extentions/google-tasks-checker-plus/src/external.png
/product/chrome-extentions/google-tasks-checker-plus/src/icon-32.png
/product/chrome-extentions/google-tasks-checker-plus/src/icon-128.png
/product/chrome-extentions/google-tasks-checker-plus/src/icon.gif
/product/chrome-extentions/google-tasks-checker-plus/src/jquery-1.4.2.min.js
/product/chrome-extentions/google-tasks-checker-plus/src/main.js
/product/chrome-extentions/google-tasks-checker-plus/src/manifest.json
/product/chrome-extentions/google-tasks-checker-plus/src/options.html
/product/chrome-extentions/google-tasks-checker-plus/src/popup.html
/product/chrome-extentions/google-tasks-checker-plus/src/reset.css
```

### printf option ( -p )

see also list command's printf option.

## du

display disk usage statistics.

- syntax

    dropbox-api du &lt;dropbox\_path> \[options\]

### Example

```
> dropbox-api du /product -h -d 1
1.1M    /product
1.1M    /product/chrome-extensions
  0B    /product/work
```

### human readable option ( -h )

print sizes in human readable format (e.g., 1K 234M 2G)

### depth option ( -d )

Display an entry for all files and directories depth directories deep.

## sync ( rsync )

recursive file synchronization.

### sync from dropbox

dropbox-api sync dropbox:&lt;source\_dir> &lt;target\_dir> \[options\]

```
> dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product
download /private/tmp/product/external.png
download /private/tmp/product/icon-32.png
download /private/tmp/product/icon-128.png
```

### sync to dropbox

dropbox-api sync &lt;source\_dir> dropbox:&lt;target\_dir> \[options\]

```
> dropbox-api sync /tmp/product dropbox:/work/src
upload background.html /work/src/background.html
upload external.png /work/src/external.png
upload icon-128.png /work/src/icon-128.png
```

### delete option ( -d )

```
> dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -d
download /private/tmp/product/external.png
download /private/tmp/product/icon-32.png
download /private/tmp/product/icon-128.png
remove background.html.tmp
```

### dry run option ( -n )

```
> dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -dn
!! enable dry run !!
download /private/tmp/product/external.png
download /private/tmp/product/icon-32.png
download /private/tmp/product/icon-128.png
remove background.html.tmp
```

### verbose option ( -v )

```
> dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -dnv
remote_base: /product/chrome-extentions/google-tasks-checker-plus/src
local_base: /private/tmp/product
** download **
skip background.html
download /private/tmp/product/external.png
download /private/tmp/product/icon-32.png
download /private/tmp/product/icon-128.png
skip icon.gif
skip jquery-1.4.2.min.js
skip main.js
skip manifest.json
skip options.html
skip popup.html
skip reset.css
** delete **
skip background.html
remove background.html.tmp
skip icon.gif
skip jquery-1.4.2.min.js
skip main.js
skip manifest.json
skip options.html
skip popup.html
skip reset.css
```

## cp

copy file or directory.

- alias

    copy

- syntax

    dropbox-api cp &lt;source\_file> &lt;target\_file>

### Example

```
dropbox-api cp memo.txt memo.txt.bak
```

## mv

move file or directory.

- alias

    move

- syntax

    dropbox-api mv &lt;source\_file> &lt;target\_file>

### Example

```
dropbox-api mv memo.txt memo.txt.bak
```

## mkdir

make directory.

\*no error if existing, make parent directories as needed.\*

- alias

    mkpath

- syntax

    dropbox-api mkdir &lt;directory>

### Example

```
dropbox-api mkdir product/src
```

## rm

remove file or directory.

\*remove the contents of directories recursively.\*

- alias

    rmtree

- syntax

    dropbox-api rm &lt;file\_or\_directory>

### Example

```
dropbox-api rm product/src
```

## get

download file from dropbox.

- alias

    dl, download

- syntax

    dropbox-api get dropbox:&lt;dropbox\_file> &lt;file>

### Example

```
dropbox-api get dropbox:/Public/foo.txt /tmp/foo.txt
```

## put

upload file to dropbox.

- alias

    up, upload

- syntax

    dropbox-api put &lt;file> dropbox:&lt;dropbox\_dir>

### Example

```
dropbox-api put /tmp/foo.txt dropbox:/Public/
```

### verbose option ( -v )

A progress bar is displayed.

```perl
dropbox-api put /tmp/1GB.dat dropbox:/Public/ -v
100% [=====================================================================================>]
```

## Tips

### Retry

```
#!/bin/bash

command='dropbox-api sync dropbox:/test/ /Users/aska/test/ -vde'

NEXT_WAIT_TIME=0
EXIT_CODE=0
until $command || [ $NEXT_WAIT_TIME -eq 4 ]; do
    EXIT_CODE=$?
    sleep $NEXT_WAIT_TIME
    let NEXT_WAIT_TIME=NEXT_WAIT_TIME+1
done
exit $EXIT_CODE
```

# COPYRIGHT

Copyright 2012- Shinichiro Aska

The standalone executable contains the following modules embedded.

# LICENSE

Released under the MIT license. http://creativecommons.org/licenses/MIT/

# COMMUNITY

- [https://github.com/s-aska/dropbox-api-command](https://github.com/s-aska/dropbox-api-command) - source code repository, issue tracker


Changes
Revision history for dropbox-api

{{$NEXT}}

2.13 2017-05-31T11:33:47Z
        - Update module version

2.12 2017-02-07T09:23:46Z
        - fix typo
        - use upload_session

2.11 2017-01-01T18:58:11Z
        - parse_datetime is %Y-%m-%dT%TZ
        - format_datetime is %Y-%m-%dT%T

2.10 2017-01-01T15:58:01Z
        - Add Z for strformat

2.09 2016-07-18T10:12:03Z
        - Fix bug `sync dropbox: ~/Dropbox`

2.08 2016-07-18T10:00:42Z
        - Fix bug `sync dropbox:/ ~/Dropbox/`

2.07 2016-07-14T12:12:15Z
        - Add sync file `sync dropbox:/file.txt ~/Dropbox/` `sync ~/file.txt dropbox:/Dropbox/`

2.06 2016-07-14T08:12:56Z
        - Fix bug `sync dropbox:/ ~/Dropbox/`

2.05 2016-07-12T11:58:23Z
        - Fix bug path_lower has upper case

2.04 2016-07-12T11:30:53Z
        - Add `du` command

2.03 2016-07-11T16:13:38Z
        - Support `ls /` and `find /`

2.02 2016-07-11T15:00:38Z
        - Update cpanfile

2.01 2016-07-11T13:17:43Z
        - Fix typo and misspell.
        - Add `-s` option documented.
        - Fix setup example.
        Thanks mathstuf

2.00 2016-07-11T11:06:31Z
        - Migration API v2

1.17    Jun 18, 2014
        - fix bug `next` => `return`

1.16    Jun 10, 2014
        - fix bug `next` => `return`

1.15    Feb 18, 2013
        - catch open error

1.14    Oct  2, 2013
        - added LICENSE, README.md, cpanfile

1.13    Jul  8, 2013
        - Remove interpolation in printf format arguments. (thefloweringash)
        - Ignore stty errors when stdin isn't a terminal. (thefloweringash)

1.12    Jun 21, 2013
        - Dropbox are case insensitive. (zenoamaro)

1.11    Jun 20, 2013
        - Dropbox are case insensitive.

1.10    Jun 20, 2013
        - Use lowercase for HTTP_PROXY envvar (mathstuf)

1.09    Jan  8, 2013
        - fix "list" and "find" utf8 path bug.

1.08    Jan  8, 2013
        - added print option. ( %R ... rev, %c %Ck ... client_mtime )
        - fix print option bug.

1.07    Jan  8, 2013
        - binmode STDERR, ":utf8"

1.06    Jan  8, 2013
        - added -e option, it is "HTTP_PROXY" and "NO_PROXY" environment variable support.
        - fix UTF-8-mac filename bug

1.05    Aug 28, 2012
        - added ProgressBar. ( need -v option )

1.04    Aug 27, 2012
        - large file support. ( over 150 MB )

1.03    Aug 14, 2012
        - set modification time.
        - support windows homepath.

1.02    Jan 10, 2012
        - added Exception handling

1.01    Apr 5, 2012
        - added sub command uid (sylvainfilteau)
        - added upload-to-dropbox command (sylvainfilteau)


sync.pl
#!perl

use strict;
use Encode;
use JSON::XS;
use File::Spec::Functions;
use File::Basename qw(dirname);
use Net::Dropbox::API;
use Path::Class;
use Getopt::Std;

my %opts;
getopts('ndf:t:', \%opts);

my $dry    = $opts{n};
my $delete = $opts{d};
my $base   = $opts{f};
my $dir    = $opts{t};
warn "enable dry mode" if $dry;
warn "!!! enable delete mode !!!" if $delete;
die "missing dest dir $dir" if !-d $dir;
$dir = dir($dir)->absolute->resolve;
my $cache_file = catfile( dirname(__FILE__), '.dropbox.cache' );
my $cache = -f $cache_file ? decode_json(file($cache_file)->slurp) : {};
die $@ if $@;
my $config_file = catfile( dirname(__FILE__), $ENV{'CONFIG_PATH'} || 'config.pl' );
my $config = do $config_file or die "load error $config_file";
$config->{key} or die "please set config key.";
$config->{secret} or die "please set config secret.";
$config->{access_token} or die "please set config access_token.";
$config->{access_secret} or die "please set config access_secret.";
my $box = Net::Dropbox::API->new($config);
$box->context('dropbox');

my $files;
my $find;
$find = sub {
    my $path = shift;
    $path = substr $path, 1 if substr($path, 0, 1) eq '/';
    #warn "find $path";
    my $list = $box->list($path);
    my @dirs;
    for my $content (@{$list->{contents}}) {
        #warn "content $content->{path}";
        my $l_path = $content->{path};
        $l_path=~s|^$base/?||;
        #warn "push $l_path";
        $files->{$l_path} = $content;
        if ($content->{is_dir}) {
            push @dirs, $content;
            my $cur = dir($dir, $l_path);
            if (!-d $cur) {
                warn "mkpath $cur";
                next if $dry;
                $cur->mkpath;
            }
        } else {
            if (
                exists $cache->{$l_path} &&
                -f file($dir, $l_path) &&
                $cache->{$l_path}->{modified} eq $files->{$l_path}->{modified}
            ) {
                warn "skip $content->{path}";
            } else {
                warn "download $content->{path}";
                next if $dry;
                my $file = file($dir, $l_path);
                $file->dir->mkpath if !-d $file->dir;
                $box->getfile(substr($content->{path}, 1), $file->absolute->stringify);
            }
        }
    }
    for my $content (@dirs) {
        $find->($content->{path}.'/');
    }
};
$find->($base);
if (not $dry) {
    my $fh = file($cache_file)->openw;
    $fh->print(encode_json($files));
    $fh->close;
}
exit if not $delete;
$dir->recurse(
    preorder => 0,
    depthfirst => 1,
    callback => sub {
        my $file = shift;
        my $path = $file->absolute;
        $path=~s|$dir/?||;
        return if not length $path;
        #warn $path;
        if (not exists $files->{decode('utf8', $path)}) {
            if (-f $file) {
                warn "unlink " . $file;
                return if $dry;
                unlink $file;
            } elsif (-d $file) {
                warn "rmtree " . $file;
                return if $dry;
                $file->rmtree;
            }
        }
    }
);

exit(0);


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'App-dropboxapi',
    module_name     => 'App::dropboxapi',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


Makefile.PL
if (-e 'dropbox-api') {
    print STDERR "Generating bin/dropbox-api from dropbox-api\n";
    mkdir 'bin' unless -d 'bin';
    open my $in,  "<dropbox-api"     or die $!;
    open my $out, ">bin/dropbox-api" or die $!;
    while (<$in>) {
        s|^#!/usr/bin/env perl|#!perl|; # so MakeMaker can fix it
        print $out $_;
    }
}
use ExtUtils::MakeMaker;
WriteMakefile(
    NAME => 'App::dropboxapi',
    VERSION_FROM => 'lib/App/dropboxapi.pm',
    ABSTRACT => 'Dropbox API Wrapper Command.',
    PREREQ_PM => {
        'ExtUtils::MakeMaker' => 6.30,
        'ExtUtils::Install' => 1.43,
        'JSON' => 0,
        'Path::Class' => 0.26,
        'WebService::Dropbox' => 1.17,
        'DateTime::Format::Strptime' => 0,
        'Encode::Locale' => 0,
        ($^O eq 'darwin' ? ('Encode::UTF8Mac' => 0) : ()),
    },
    LICENSE => 'mit',
    EXE_FILES => [ 'bin/dropbox-api' ],
);

dropbox-setup.pl
#!perl

use strict;
use File::Basename qw(dirname);
use JSON;
use Net::Dropbox::API;
use Path::Class;
my $config_file = file( dirname(__FILE__), $ENV{'CONFIG_PATH'} || 'dropbox-config.json' );
my $config = {};

print "Please Input API Key: ";
chomp( my $key = <STDIN> );
die 'Get API Key from https://www.dropbox.com/developers' unless $key;
$config->{key} = $key;

print "Please Input API Secret: ";
chomp( my $secret = <STDIN> );
die 'Get API Secret from https://www.dropbox.com/developers' unless $secret;
$config->{secret} = $secret;

$config->{callback_url} = '';
my $box = Net::Dropbox::API->new($config);
my $login_link = $box->login;
print "URL: $login_link\n";
print "Please Access URL and AUTH\n";
print "OK?";
<STDIN>;
$box->auth;
$config->{access_token} = $box->access_token;
$config->{access_secret} = $box->access_secret;
print "success! try\n";
print "> perl dropbox-api.pl ls\n";
print "> perl dropbox-api.pl find /\n";

$config_file->openw->print(encode_json($config));

exit(0);


lib/App/dropboxapi.pm
package App::dropboxapi;
use strict;
use warnings;
our $VERSION = '2.13';

=head1 NAME

App::dropboxapi - command line interface to access Dropbox API

=head1 SYNOPSIS

    dropbox-api put /tmp/foo.txt dropbox:/Public/

Run C<dropbox-api help> for more options.

=head1 DESCRIPTION

dropbox-api is a command line interface to access Dropbox API.

=over 4

=item ls

=item find

=item du

=item sync

=item cp

=item mv

=item rm

=item mkdir

=item get

=item put

=back

=head1 Install and Setup

=head2 1. Install

=head3 1-a) FreeBSD

    pkg_add -r dropbox-api-command

=head3 1-b) Ubuntu

    sudo apt-get install make gcc libssl-dev wget
    wget https://raw.github.com/miyagawa/cpanminus/master/cpanm
    sudo perl cpanm App::dropboxapi

=head3 1-c) CentOS

    # CentOS 5
    sudo yum install gcc gcc-c++ openssl-devel wget
    # CentOS 6
    sudo yum install gcc gcc-c++ openssl-devel wget perl-devel
    wget https://raw.github.com/miyagawa/cpanminus/master/cpanm
    sudo perl cpanm App::dropboxapi

=head3 1-d) OS X

    # Install Command Line Tools for Xcode
    open https://www.google.com/search?q=Command+Line+Tools+for+Xcode

    curl -O https://raw.github.com/miyagawa/cpanminus/master/cpanm
    sudo perl cpanm App::dropboxapi

=head2 2. Get API Key and API Secret

    https://www.dropbox.com/developers
    My Apps => Create an App

=head2 3. Get Access Token and Access Secret

    > dropbox-api setup
    Please Input API Key: ***************
    Please Input API Secret: ***************
    1. Open the Login URL: https://www.dropbox.com/oauth2/authorize?client_id=*****&response_type=code
    2. Input code and press Enter: ***************
    success! try
    > dropbox-api ls
    > dropbox-api find /

=head2 4. How to use Proxy

Please use -e option.

    > HTTP_PROXY="http://127.0.0.1:8888" dropbox-api setup -e

=head1 Sub Commands

=head2 help

disp help.

=over 4

=item syntax

dropbox-api help [<command>]

=back

=head3 Example

    > dropbox-api help
    Usage: dropbox-api <command> [args] [options]

    Available commands:
        setup get access_key and access_secret
        ls    list directory contents
        find  walk a file hierarchy
        du    disk usage statistics
        cp    copy file or directory
        mv    move file or directory
        mkdir make directory (Create intermediate directories as required)
        rm    remove file or directory (Attempt to remove the file hierarchy rooted in each file argument)
        put   upload file
        get   download file
        sync  sync directory (local => dropbox or dropbox => local)

    Common Options
        -e enable env_proxy ( HTTP_PROXY, NO_PROXY )
        -D enable debug
        -v verbose
        -s sandbox mode, but this option has been removed.

    See 'dropbox-api help <command>' for more information on a specific command.

=head3 Example ( command help )

    > dropbox-api help ls
    Name
        dropbox-api-ls - list directory contents

    SYNOPSIS
        dropbox-api ls <dropbox_path> [options]

    Example
        dropbox-api ls Public
        dropbox-api ls Public -h
        dropbox-api ls Public -p "%d\t%s\t%TY/%Tm/%Td %TH:%TM:%TS\t%p\n"

    Options
        -h print sizes in human readable format (e.g., 1K 234M 2G)
        -p print format.
            %d ... is_dir ( d: dir, -: file )
            %i ... id
            %n ... name
            %p ... path_display
            %P ... path_lower
            %b ... bytes
            %s ... size (e.g., 1K 234M 2G)
            %t ... server_modified
            %c ... client_modified
            %r ... rev
            %Tk ... DateTime 'strftime' function (server_modified)
            %Ck ... DateTime 'strftime' function (client_modified)

L<http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime_Patterns>

=head2 ls

file list view.

=over 4

=item alias

list

=item syntax

dropbox-api ls <dropbox_path>

=back

=head3 Example

    > dropbox-api list /product
    d        - Thu, 24 Feb 2011 06:58:00 +0000 /product/chrome-extentions
    -   294557 Sun, 26 Dec 2010 21:55:59 +0000 /product/ex.zip

=head3 human readable option ( -h )

print sizes in human readable format (e.g., 1K 234M 2G)

    > dropbox-api ls /product -h
    d        - Thu, 24 Feb 2011 06:58:00 +0000 /product/chrome-extentions
    -  287.7KB Sun, 26 Dec 2010 21:55:59 +0000 /product/ex.zip

=head3 printf option ( -p )

print format.

    > dropbox-api ls /product -p "%d\t%s\t%TY/%Tm/%Td %TH:%TM:%TS\t%p\n"
    d       -       2011/02/24 06:58:00     /product/chrome-extentions
    -       287.7KB 2010/12/26 21:55:59     /product/ex.zip

        %d ... is_dir ( d: dir, -: file )
        %i ... id
        %n ... name
        %p ... path_display
        %P ... path_lower
        %b ... bytes
        %s ... size (e.g., 1K 234M 2G)
        %t ... server_modified
        %c ... client_modified
        %r ... rev
        %Tk ... DateTime 'strftime' function (server_modified)
        %Ck ... DateTime 'strftime' function (client_modified)

L<http://search.cpan.org/dist/DateTime/lib/DateTime.pm#strftime_Patterns>

=head2 find

recursive file list view.

=over 4

=item syntax

dropbox-api find <dropbox_path> [options]

=back

=head3 Example

    > dropbox-api find /product/google-tasks-checker-plus
    /product/chrome-extentions/google-tasks-checker-plus/README.md
    /product/chrome-extentions/google-tasks-checker-plus/src
    /product/chrome-extentions/google-tasks-checker-plus/src/background.html
    /product/chrome-extentions/google-tasks-checker-plus/src/external.png
    /product/chrome-extentions/google-tasks-checker-plus/src/icon-32.png
    /product/chrome-extentions/google-tasks-checker-plus/src/icon-128.png
    /product/chrome-extentions/google-tasks-checker-plus/src/icon.gif
    /product/chrome-extentions/google-tasks-checker-plus/src/jquery-1.4.2.min.js
    /product/chrome-extentions/google-tasks-checker-plus/src/main.js
    /product/chrome-extentions/google-tasks-checker-plus/src/manifest.json
    /product/chrome-extentions/google-tasks-checker-plus/src/options.html
    /product/chrome-extentions/google-tasks-checker-plus/src/popup.html
    /product/chrome-extentions/google-tasks-checker-plus/src/reset.css

=head3 printf option ( -p )

see also list command's printf option.

=head2 du

display disk usage statistics.

=over 4

=item syntax

dropbox-api du <dropbox_path> [options]

=back

=head3 Example

    > dropbox-api du /product -h -d 1
    1.1M    /product
    1.1M    /product/chrome-extensions
      0B    /product/work

=head3 human readable option ( -h )

print sizes in human readable format (e.g., 1K 234M 2G)

=head3 depth option ( -d )

Display an entry for all files and directories depth directories deep.

=head2 sync ( rsync )

recursive file synchronization.

=head3 sync from dropbox

dropbox-api sync dropbox:<source_dir> <target_dir> [options]

    > dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product
    download /private/tmp/product/external.png
    download /private/tmp/product/icon-32.png
    download /private/tmp/product/icon-128.png

=head3 sync to dropbox

dropbox-api sync <source_dir> dropbox:<target_dir> [options]

    > dropbox-api sync /tmp/product dropbox:/work/src
    upload background.html /work/src/background.html
    upload external.png /work/src/external.png
    upload icon-128.png /work/src/icon-128.png

=head3 delete option ( -d )

    > dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -d
    download /private/tmp/product/external.png
    download /private/tmp/product/icon-32.png
    download /private/tmp/product/icon-128.png
    remove background.html.tmp

=head3 dry run option ( -n )

    > dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -dn
    !! enable dry run !!
    download /private/tmp/product/external.png
    download /private/tmp/product/icon-32.png
    download /private/tmp/product/icon-128.png
    remove background.html.tmp

=head3 verbose option ( -v )

    > dropbox-api sync dropbox:/product/google-tasks-checker-plus/src /tmp/product -dnv
    remote_base: /product/chrome-extentions/google-tasks-checker-plus/src
    local_base: /private/tmp/product
    ** download **
    skip background.html
    download /private/tmp/product/external.png
    download /private/tmp/product/icon-32.png
    download /private/tmp/product/icon-128.png
    skip icon.gif
    skip jquery-1.4.2.min.js
    skip main.js
    skip manifest.json
    skip options.html
    skip popup.html
    skip reset.css
    ** delete **
    skip background.html
    remove background.html.tmp
    skip icon.gif
    skip jquery-1.4.2.min.js
    skip main.js
    skip manifest.json
    skip options.html
    skip popup.html
    skip reset.css

=head2 cp

copy file or directory.

=over 4

=item alias

copy

=item syntax

dropbox-api cp <source_file> <target_file>

=back

=head3 Example

    dropbox-api cp memo.txt memo.txt.bak

=head2 mv

move file or directory.

=over 4

=item alias

move

=item syntax

dropbox-api mv <source_file> <target_file>

=back

=head3 Example

    dropbox-api mv memo.txt memo.txt.bak

=head2 mkdir

make directory.

*no error if existing, make parent directories as needed.*

=over 4

=item alias

mkpath

=item syntax

dropbox-api mkdir <directory>

=back

=head3 Example

    dropbox-api mkdir product/src

=head2 rm

remove file or directory.

*remove the contents of directories recursively.*

=over 4

=item alias

rmtree

=item syntax

dropbox-api rm <file_or_directory>

=back

=head3 Example

    dropbox-api rm product/src

=head2 get

download file from dropbox.

=over 4

=item alias

dl, download

=item syntax

dropbox-api get dropbox:<dropbox_file> <file>

=back

=head3 Example

    dropbox-api get dropbox:/Public/foo.txt /tmp/foo.txt

=head2 put

upload file to dropbox.

=over 4

=item alias

up, upload

=item syntax

dropbox-api put <file> dropbox:<dropbox_dir>

=back

=head3 Example

    dropbox-api put /tmp/foo.txt dropbox:/Public/

=head3 verbose option ( -v )

A progress bar is displayed.

    dropbox-api put /tmp/1GB.dat dropbox:/Public/ -v
    100% [=====================================================================================>]

=head2 Tips

=head3 Retry

    #!/bin/bash

    command='dropbox-api sync dropbox:/test/ /Users/aska/test/ -vde'

    NEXT_WAIT_TIME=0
    EXIT_CODE=0
    until $command || [ $NEXT_WAIT_TIME -eq 4 ]; do
        EXIT_CODE=$?
        sleep $NEXT_WAIT_TIME
        let NEXT_WAIT_TIME=NEXT_WAIT_TIME+1
    done
    exit $EXIT_CODE

=head1 COPYRIGHT

Copyright 2012- Shinichiro Aska

The standalone executable contains the following modules embedded.

=head1 LICENSE

Released under the MIT license. http://creativecommons.org/licenses/MIT/

=head1 COMMUNITY

=over 4

=item L<https://github.com/s-aska/dropbox-api-command> - source code repository, issue tracker

=back

=cut

1;
